Source/Container/HCMap.c

```diff
@@ -294,11 +293,17 @@ HCBoolean HCMapIterationHasBegun(HCMapIterator* iterator) {
 
 HCBoolean HCMapIterationHasNext(HCMapIterator* iterator) {
     HCSetIterator* pairIterator = iterator->state;
-    return HCSetIterationHasNext(pairIterator);
+    return pairIterator == NULL ? false:HCSetIterationHasNext(pairIterator);
 }
 
 HCBoolean HCMapIterationHasEnded(HCMapIterator* iterator) {
-    return HCMapIterationHasBegun(iterator) && iterator->map != NULL && (iterator->key == NULL || iterator->object == NULL);
+    if (HCMapIterationHasBegun(iterator) && iterator->map != NULL && (iterator->key == NULL || iterator->object == NULL)) {
+        if (iterator->state != NULL) {
+            HCMapIterationEnd(iterator);
```

Comment:
This function should not have side effects, should it?

Reply:
So after looking at it I can move the clean up to `HCMapIterationNext()` but we should probably have the clean up code in a consistent place in all of the iterators so if they change implementation requirements (for class clusters or some other reason) they will just clean themselves up.

The potential issue I am running into is with `HCListIterationNext()` it breaks certain tests since it goes to a dead state as soon as it goes off the end. Is this acceptable or do we want to be able to go off the end on `HCList` and then go back one step or for consistency should we not allow this either?